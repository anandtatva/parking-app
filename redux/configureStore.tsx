import { persistReducer, persistStore } from "redux-persist";
import { combineReducers, createStore } from "redux";
import parkingSlotSlice from "./reducer/parking-slot-slice";
import AsyncStorage from "@react-native-async-storage/async-storage";


const persistConfig = {
    key: "parking-slot",
    storage: AsyncStorage
}

const rootReducers = combineReducers({
    parkingSlot: parkingSlotSlice
})

const slotPersistReducer = persistReducer(persistConfig, rootReducers);

const store = createStore(slotPersistReducer);

const persistor = persistStore(store)

export { store, persistor }