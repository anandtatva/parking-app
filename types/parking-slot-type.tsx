type ParkingSlotType = {
    id: string;
    slotNumber: number;
    isEmpty: boolean;
    carNumber: string;
    enterTime: null | number;
    exitTime: null | number;
}

export type {
    ParkingSlotType
}