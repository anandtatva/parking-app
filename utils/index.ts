export * from './strings';
export * from './colors';
export * from "./utils";
export * from "./alerts";