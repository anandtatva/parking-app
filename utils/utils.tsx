import moment from "moment"
function validateCarNumber(carNumber: string) {
    const pattern = /[A-Z][A-Z] [0-9][0-9] [A-Z][A-Z] [0-9][0-9][0-9][0-9]$/i;
    const re = new RegExp(pattern);
    return re.test(carNumber)
}
function msToTime(time: number) {
    return moment(time * 1000).format("DD/MM/YYYY HH:mm:ss");

}
function getDifferenceDateString(startTime: number) {
    const { hour, minutes } = getDifferenceDate(startTime, true)
    if (hour && minutes)
        return `${hour} hours ${minutes} minutes`
    else if (hour) return `${hour} hour`
    else if (minutes) return `${minutes} minutes`
    else return `${hour} hours`
}
function getDifferenceDate(enterTime: number, status = false) {
    const totalMinutes = moment().diff(moment(enterTime * 1000), "minutes")
    if (!isNaN(totalMinutes) && totalMinutes) {
        const hours = Math.floor(totalMinutes / 60)
        const minutes = totalMinutes % 60;
        if (status)
            return { hour: hours, minutes }
        else
            return { hour: hours + (minutes > 0 && 1 || 0), minutes }
    } else return { hour: 1, minutes: 0 }

}
export {
    validateCarNumber,
    msToTime,
    getDifferenceDateString,
    getDifferenceDate
}