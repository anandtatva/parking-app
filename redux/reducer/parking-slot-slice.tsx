import { createSlice } from "@reduxjs/toolkit";
import { ParkingSlotType } from "../../types/parking-slot-type";

interface StateProps {
    parkingSlots: ParkingSlotType[];
}
const initialState: StateProps = {
    parkingSlots: []
}

const parkingSlotSlice = createSlice({
    name: "parking-slot",
    initialState,
    reducers: {
        setParkingSlot: (state, action) => {
            state.parkingSlots = action.payload;
        }
    }
})

export const { setParkingSlot } = parkingSlotSlice.actions;

export default parkingSlotSlice.reducer;