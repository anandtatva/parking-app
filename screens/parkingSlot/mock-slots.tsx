import { ParkingSlotType } from "../../types/parking-slot-type";

const mockSlots: ParkingSlotType[] = [
    {
        id: '55bfe613-f4b1-4bbf-8ebf-f8a661b7519c',
        slotNumber: 1,
        isEmpty: true,
        carNumber: '',
        enterTime: null,
        exitTime: null
    },
    {
        id: '81075ab6-823e-44b3-8015-d034df652950',
        slotNumber: 2,
        isEmpty: true,
        carNumber: '',
        enterTime: null,
        exitTime: null
    }
]

const fullMockSlot: ParkingSlotType[] = [
    {
        id: '55bfe613-f4b1-4bbf-8ebf-f8a661b7519c',
        slotNumber: 1,
        isEmpty: false,
        carNumber: 'XZ 08 XZ 0000',
        enterTime: 1656554945,
        exitTime: null
    },
    {
        id: '81075ab6-823e-44b3-8015-d034df652950',
        slotNumber: 2,
        isEmpty: false,
        carNumber: 'XZ 08 XZ 0000',
        enterTime: 1656564945,
        exitTime: null
    }
]

export { mockSlots, fullMockSlot };