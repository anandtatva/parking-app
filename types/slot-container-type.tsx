import { ParkingSlotType } from "./parking-slot-type";

export interface SlotContainerType {
    item: ParkingSlotType,
    index: number
    onRemoveCar: (index: number, item: ParkingSlotType) => void
}