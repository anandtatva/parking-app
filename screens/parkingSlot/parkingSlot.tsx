import { StatusBar } from 'expo-status-bar';
import { Platform, Alert, Text, StyleSheet, View, FlatList } from 'react-native';

import EditScreenInfo from '../../components/EditScreenInfo';
import { useSelector, useDispatch } from "react-redux";
import { SlotContainer } from './slot-container';
import { ParkingSlotType } from '../../types/parking-slot-type';
import { string } from '../../utils';
import moment from "moment";
import { setParkingSlot } from '../../redux/reducer/parking-slot-slice';

export default function ParkingSlot({ navigation }) {
  const dispatch = useDispatch()
  const parkingSlots = useSelector(state => state.parkingSlot)
  const removeCarAlert = (index: number, item: ParkingSlotType) => {
    if (Platform.OS === "web") {
      const resp = confirm("Are you sure you want to remove car from parking?");
      resp && removeCar(index, item);

    } else {
      Alert.alert('', 'Are you sure you want to remove car from parking?', [
        { text: 'Cancel', onPress: () => null },
        { text: 'OK', onPress: () => removeCar(index, item) },
      ], { cancelable: false })
    }

  }

  const removeCar = (index: number, item: ParkingSlotType) => {
    let updateCar: ParkingSlotType = { ...parkingSlots.parkingSlots[index] };
    updateCar.exitTime = moment().unix()
    const tempList = JSON.parse(JSON.stringify(parkingSlots.parkingSlots));
    tempList.splice(index, 1, updateCar)
    dispatch(setParkingSlot(tempList))
    navigation.navigate("PaymentScreen", { slotDetail: updateCar })
  }
  const renderItem = ({ item, index }) => (
    <SlotContainer
      item={item}
      index={index}
      onRemoveCar={removeCarAlert}
    />
  )
  return (
    <View style={styles.container}>
      {parkingSlots?.parkingSlots?.length ? (
        <Text style={styles.helpTxt}>{string.noteRemoveCar}</Text>
      ) : null}

      <FlatList
        testID="Slot-List"
        numColumns={2}
        keyExtractor={(_, index) => String(index)}
        data={parkingSlots?.parkingSlots || []}
        renderItem={renderItem}
        ListEmptyComponent={() => (
          <Text style={styles.title} testID={"Empty-Slots"}>{string.emptySlot}</Text>
        )}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  helpTxt: {
    opacity: 0.8,
    padding: 10,
    textAlign: "center"
  },
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  title: {
    fontSize: 20,
    fontWeight: 'bold',
  },
  separator: {
    marginVertical: 30,
    height: 1,
    width: '80%',
  },
});
