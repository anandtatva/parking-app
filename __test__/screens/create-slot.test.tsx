import { render, waitFor, act, fireEvent, cleanup } from "@testing-library/react-native";
import CreateSlot from "../../screens/createSlot/createSlot";
import { Provider } from "react-redux";
import { store } from "../../redux/configureStore";
import { setParkingSlot } from "../../redux/reducer/parking-slot-slice";
import { mockSlots, fullMockSlot } from "../../screens/parkingSlot/mock-slots";
import * as Utils from "../../utils/utils";
import { Alert } from "react-native";
import { string } from "../../utils";
import { ParkingSlotType } from "../../types/parking-slot-type";

jest.mock("@react-navigation/native", () => {
    const actualNav = jest.requireActual("@react-navigation/native");
    return {
        ...actualNav,
        useFocusEffect: () => jest.fn(),
        useNavigation: () => ({
            navigate: jest.fn()
        }),
        useRoute: () => ({
            params: {}
        })
    }

})
const spyAlert = jest.spyOn(Alert, 'alert');

describe("Create Parking Slot", () => {
    afterEach(() => {
        jest.clearAllMocks();
        cleanup()
    })
    beforeEach(() => {
        store.dispatch(setParkingSlot([]))
    })
    it("renders correctly", () => {
        const navigation = {
            navigate: jest.fn()
        }
        const tree = render(
            <Provider store={store}>
                <CreateSlot navigation={navigation} />
            </Provider>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    })
    it("Vaild Number (5): User should be able to enter slot number", async () => {
        const navigation = {
            navigate: jest.fn()
        }
        const SLOT_NUMBER = "5";
        const { getByTestId } = render(
            <Provider store={store}>
                <CreateSlot navigation={navigation} />
            </Provider>
        )
        const slotNoInput = getByTestId("Slot-Number-Input");
        await waitFor(() => {
            fireEvent.changeText(slotNoInput, SLOT_NUMBER);
            expect(slotNoInput.props.value).toBe(SLOT_NUMBER)
        })
    })
    it("InVaild Number (a): User should not be able to enter slot number", async () => {
        const navigation = {
            navigate: jest.fn()
        }
        const SLOT_NUMBER = "a";
        const { getByTestId } = render(
            <Provider store={store}>
                <CreateSlot navigation={navigation} />
            </Provider>
        )
        const slotNoInput = getByTestId("Slot-Number-Input");
        await waitFor(() => {
            fireEvent.changeText(slotNoInput, SLOT_NUMBER);
            expect(slotNoInput.props.value).toBe("0")
        })
    })
    it("Create Slots: On press create slot button", async () => {
        const navigation = {
            navigate: jest.fn()
        }
        const SLOT_NUMBER = "5";
        const { getByTestId } = render(
            <Provider store={store}>
                <CreateSlot navigation={navigation} />
            </Provider>
        )
        const slotNoInput = getByTestId("Slot-Number-Input");
        const createSlotButton = getByTestId("Create-Slot-Button");
        await waitFor(() => {
            fireEvent.changeText(slotNoInput, SLOT_NUMBER);
        })
        fireEvent.press(createSlotButton)
        expect(store.getState().parkingSlot.parkingSlots.length).toBe(5)
    })
    it("Reset Slot: On press reset slot button", async () => {
        const navigation = {
            navigate: jest.fn()
        }
        const SLOT_NUMBER = "5";
        const { getByTestId } = render(
            <Provider store={store}>
                <CreateSlot navigation={navigation} />
            </Provider>
        )
        const slotNoInput = getByTestId("Slot-Number-Input");
        const createSlotButton = getByTestId("Create-Slot-Button");
        await waitFor(() => {
            fireEvent.changeText(slotNoInput, SLOT_NUMBER);
        })
        fireEvent.press(createSlotButton)
        expect(store.getState().parkingSlot.parkingSlots.length).toBe(5)
        const resetSlotButton = getByTestId("Reset-Slot-Button");
        fireEvent.press(resetSlotButton)
        expect(store.getState().parkingSlot.parkingSlots.length).toBe(0)
    })
    it("View Slot: On press view slot button", async () => {
        const navigation = {
            navigate: jest.fn()
        }
        const { getByTestId } = render(
            <Provider store={store}>
                <CreateSlot navigation={navigation} />
            </Provider>
        )
        await act(async () => {
            const viewSlotButton = getByTestId("View-Slot-Button");
            store.dispatch(setParkingSlot(mockSlots))
            expect(store.getState().parkingSlot.parkingSlots.length).toBe(2)
            fireEvent.press(viewSlotButton)
            expect(navigation.navigate).toHaveBeenCalledWith('ParkingSlotScreen');
        })
    })
    it("View Slot: On press view slot button", async () => {
        const navigation = {
            navigate: jest.fn()
        }
        const { getByTestId } = render(
            <Provider store={store}>
                <CreateSlot navigation={navigation} />
            </Provider>
        )
        await act(async () => {
            const viewSlotButton = getByTestId("View-Slot-Button");
            store.dispatch(setParkingSlot(mockSlots))
            expect(store.getState().parkingSlot.parkingSlots.length).toBe(2)
            fireEvent.press(viewSlotButton)
            expect(navigation.navigate).toHaveBeenCalledWith('ParkingSlotScreen');
        })
    })
    it("Add Car in slot: invalid car number -> show error", async () => {
        const navigation = {
            navigate: jest.fn()
        }
        const { getByTestId } = render(
            <Provider store={store}>
                <CreateSlot navigation={navigation} />
            </Provider>
        )
        await act(async () => {
            store.dispatch(setParkingSlot(mockSlots))
            expect(store.getState().parkingSlot.parkingSlots.length).toBe(2)
            const addCarInSlotButton = getByTestId("Add-Car-In-Slot-Button");
            const addCarInSlotInput = getByTestId("Add-Car-In-Slot-Input");
            const carNo = "XZ07ZX9890";
            await waitFor(() => {
                fireEvent.changeText(addCarInSlotInput, carNo)
            })
            expect(addCarInSlotInput.props.value).toBe(carNo)
            jest.spyOn(Utils, "validateCarNumber").mockReturnValue(false)
            fireEvent.press(addCarInSlotButton);
            expect(spyAlert.mock.calls[0][0]).toBe(string.validCarError)
        })
    })
    it("Add Car in slot: slot not available -> show error", async () => {
        const navigation = {
            navigate: jest.fn()
        }
        const { getByTestId } = render(
            <Provider store={store}>
                <CreateSlot navigation={navigation} />
            </Provider>
        )
        await act(async () => {
            store.dispatch(setParkingSlot(fullMockSlot))
            expect(store.getState().parkingSlot.parkingSlots.length).toBe(2)
            const addCarInSlotButton = getByTestId("Add-Car-In-Slot-Button");
            const addCarInSlotInput = getByTestId("Add-Car-In-Slot-Input");
            const carNo = "XZ 07 ZX 9890";
            await waitFor(() => {
                fireEvent.changeText(addCarInSlotInput, carNo)
                expect(addCarInSlotInput.props.value).toBe(carNo)
                jest.spyOn(Utils, "validateCarNumber").mockReturnValue(true)
                fireEvent.press(addCarInSlotButton);
                expect(spyAlert.mock.calls[0][0]).toBe(string.parkingFull)
            })

        })
    })
    it("Add Car in slot: valid car number -> success", async () => {
        const navigation = {
            navigate: jest.fn()
        }
        const { getByTestId } = render(
            <Provider store={store}>
                <CreateSlot navigation={navigation} />
            </Provider>
        )
        await act(async () => {
            store.dispatch(setParkingSlot(mockSlots))
            expect(store.getState().parkingSlot.parkingSlots.length).toBe(2)
            const addCarInSlotButton = getByTestId("Add-Car-In-Slot-Button");
            const addCarInSlotInput = getByTestId("Add-Car-In-Slot-Input");
            const carNo = "XZ 07 ZX 9890";
            await waitFor(() => {
                fireEvent.changeText(addCarInSlotInput, carNo)
                expect(addCarInSlotInput.props.value).toBe(carNo)
                jest.spyOn(Utils, "validateCarNumber").mockReturnValue(true)
                fireEvent.press(addCarInSlotButton);
                const bookedSlot = store.getState().parkingSlot.parkingSlots.filter(({ isEmpty }: ParkingSlotType) => !isEmpty)
                expect(bookedSlot.length).toBe(1)
            })

        })
    })
})