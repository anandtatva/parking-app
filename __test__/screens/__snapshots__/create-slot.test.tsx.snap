// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Parking Slot renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "paddingVertical": 20,
    }
  }
>
  <TextInput
    keyboardType="number-pad"
    onChangeText={[Function]}
    placeholder="Enter Parking Slot Number"
    placeholderTextColor="gray"
    style={
      Object {
        "borderColor": "gray",
        "borderWidth": 1,
        "paddingHorizontal": 10,
        "paddingVertical": 5,
      }
    }
    testID="Slot-Number-Input"
    value="10"
  />
  <View
    style={
      Object {
        "flexDirection": "row",
        "marginVertical": 20,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#000",
          "opacity": 1,
          "paddingHorizontal": 8,
          "paddingVertical": 10,
        }
      }
      testID="Create-Slot-Button"
    >
      <Text
        style={
          Object {
            "color": "#fff",
            "textAlign": "center",
          }
        }
      >
        Create Slot
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#000",
          "marginHorizontal": 10,
          "opacity": 1,
          "paddingHorizontal": 8,
          "paddingVertical": 10,
        }
      }
      testID="Reset-Slot-Button"
    >
      <Text
        style={
          Object {
            "color": "#fff",
            "textAlign": "center",
          }
        }
      >
        Reset Slot
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#000",
          "opacity": 1,
          "paddingHorizontal": 8,
          "paddingVertical": 10,
        }
      }
      testID="View-Slot-Button"
    >
      <Text
        style={
          Object {
            "color": "#fff",
            "textAlign": "center",
          }
        }
      >
        View Slot
      </Text>
    </View>
  </View>
  <View>
    <TextInput
      autoCapitalize="characters"
      keyboardType="ascii-capable"
      onChangeText={[Function]}
      placeholder="Enter Car Number"
      placeholderTextColor="gray"
      style={
        Object {
          "borderColor": "gray",
          "borderWidth": 1,
          "paddingHorizontal": 10,
          "paddingVertical": 5,
        }
      }
      testID="Add-Car-In-Slot-Input"
      value="RJ 06 XP 0000"
    />
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#000",
          "marginVertical": 15,
          "opacity": 1,
          "paddingHorizontal": 8,
          "paddingVertical": 10,
        }
      }
      testID="Add-Car-In-Slot-Button"
    >
      <Text
        style={
          Object {
            "color": "#fff",
            "textAlign": "center",
          }
        }
      >
        Add Car In Parking
      </Text>
    </View>
  </View>
</View>
`;
